File: concise/conciseui/layout.cpp
#include "stdafx.h"
#include "../conciseui.h"

namespace cs{
	ICtrl* CreateCtrlObject(LPCWSTR className){
		if(WcsLength(className)==0){
			CONASSERT(L"className is null");
			return 0;
		}else if(WcsEqualNoCase(className,L"Static")){
			return (ICtrl*)new Label;
		}else if(WcsEqualNoCase(className,L"Button")){
			return new Button;
		}else if(WcsEqualNoCase(className,L"Edit")){
			return new Edit;
		}else if(WcsEqualNoCase(className,L"SysDateTimePick32")){
			return new DateTimePicker;
		}else if(WcsEqualNoCase(className,L"SysIPAddress32")){
			return new IpAddress;
		}else if(WcsEqualNoCase(className,L"SysLink")){
			return new Link;
		}else if(WcsEqualNoCase(className,L"SysHeader32")){
			return new Header;
		}else if(WcsEqualNoCase(className,L"msctls_updown32")){
			return new UpDown;
		}else if(WcsEqualNoCase(className,L"msctls_trackbar32")){
			return new TrackBar;
		}else if(WcsEqualNoCase(className,L"msctls_statusbar32")){
			return new StatusBar;
		}else if(WcsEqualNoCase(className,L"SysListView32")){
			return new ListView;
		}else if(WcsEqualNoCase(className,L"msctls_hotkey32")){
			return new HotKey;
		}else if(WcsEqualNoCase(className,L"SysTabControl32")){
			return new TabPage;
		}else if(WcsEqualNoCase(className,L"msctls_progress32")){
			return new ProgressBar;
		}else if(WcsEqualNoCase(className,L"ToolbarWindow32")){
			return new ToolBar;
		}else if(WcsEqualNoCase(className,L"SysTreeView32")){
			return new TreeView;
		}else if(WcsEqualNoCase(className,L"ComboBoxEx32")){
			return new ComboBoxEx;
		}else if(WcsEqualNoCase(className,L"ComboBox")){
			return new ComboBox;
		}else if(WcsEqualNoCase(className,L"ListBox")){
			return (ICtrl*)new ListBox;
		}else if(WcsEqualNoCase(className,L"SysMonthCal32")){
			return new MonthCalendar;
		}
		_ASSERT_EXPR(0,L"ICtrl.CreateCtrlObject not find class name");
		return new ICtrl;
	}

	void LayoutParam::SetVertical(){
		BitRemove(Style,(int)HORIZONTAL);
	}
	void LayoutParam::SetHorizontal(){
		Style |= HORIZONTAL;
	}
	void LayoutParam::Reset(){
		::ZeroMemory(this,sizeof(*this));
		Weight = 1;
	}
	void LayoutParam::SetDimen(int length,float weight){
		Length = length;
		Weight = weight;
	}
	void LayoutParam::SetMargin(int left,int top,int right,int bottom){
		Margin.SetValue(left,top,right,bottom);
	}
	LayoutParam::LayoutParam(){
		::ZeroMemory(this,sizeof(*this));
		Weight = 1;
	}
//
	inline bool weightIsZero(float weight){
		return ((weight<0.0001f)&&(weight>-0.0001f));
	}
//
	Layout::Layout(){
		_createWnd = 0;
	}
	Layout::~Layout(){
		if(_createWnd){
			_createWnd->Close();
			delete _createWnd;
		}
	}
	Twin Layout::_calcMinSize(Layout* lay,bool parentHor){
		Twin sz(0,0);
		if(parentHor){
			sz.x = lay->Param.Length;
		}else{
			sz.y = lay->Param.Length;
		}
		Twin ms(0,0);
		for(uint i=0;i<lay->_childs.Count();i++){
			Layout& child = lay->_childs[i];
			Twin childSz = _calcMinSize(&child,lay->Param.IsHorizontal());
			ms.x += childSz.x;
			ms.y += childSz.y;
		}
		if(sz.x<ms.x) sz.x = ms.x;
		if(sz.y<ms.y) sz.y = ms.y;

		sz.x += lay->Param.Margin.left + lay->Param.Margin.right;
		sz.y += lay->Param.Margin.top + lay->Param.Margin.bottom;
		return sz;
	}
	void Layout::_setRect(int left,int top,int width,int height){
		_location.x = (short)left;
		_location.y = (short)top;
		_size.x = (short)width;
		_size.y = (short)height;
		if(_wnd){
			_wnd.SetLocation(left,top);
			_wnd.SetSize(width,height);
		}
		if(Param.Style&LayoutParam::HORIZONTAL)
			_setChildHor(left,top,width,height);
		else
			_setChildVer(left,top,width,height);
	}
	void Layout::SetControl(HWND ctrl){
		if(_createWnd){
			_createWnd->Close();
			delete _createWnd;
			_createWnd = 0;
		}
		_wnd = ctrl;
	}
	void Layout::SetControl(LPCWSTR className,WORD id,LPCWSTR text,DWORD style,DWORD exstyle){
		if(WcsLength(className)!=0){
			if(_createWnd==0){
				_createWnd = CreateCtrlObject(className);
			}else{
				_createWnd->Close();
				_wnd = 0;
			}
			_createWnd->Param->ClassName = className;
			_createWnd->Param->Text = text;
			_createWnd->Param->Identity = id;
			if(style!=-1) _createWnd->Param->Style = style;
			if(exstyle!=-1) _createWnd->Param->ExStyle = exstyle;
		}else if(_createWnd){
			_createWnd->Close();
			delete _createWnd;
			_createWnd = 0;
			_wnd = 0;
		}
	}
	bool Layout::PtInLayout(int x,int y){
		Rect rect(_location.x,_location.y,_location.x+_size.x,_location.y+_size.y);
		return rect.PtInRect(x,y);
	}
	void Layout::ToJson(Json* js){
		js->SetPropertyStr(L"name",Name);
		js->SetPropertyInt32(L"style",Param.Style);
		js->SetPropertyDouble(L"weight",Param.Weight);
		js->SetPropertyInt32(L"length",Param.Length);
		js->SetPropertyInt32(L"margin_left",Param.Margin.left);
		js->SetPropertyInt32(L"margin_top",Param.Margin.top);
		js->SetPropertyInt32(L"margin_right",Param.Margin.right);
		js->SetPropertyInt32(L"margin_bottom",Param.Margin.bottom);
		if(_createWnd){
			js->SetPropertyStr(L"ctrl_class",_createWnd->Param->ClassName);
			js->SetPropertyStr(L"ctrl_text",_createWnd->Param->Text);
			js->SetPropertyInt32(L"ctrl_style",_createWnd->Param->Style);
			js->SetPropertyInt32(L"ctrl_exstyle",_createWnd->Param->ExStyle);
			js->SetPropertyInt32(L"ctrl_id",_createWnd->Param->Identity);
		}
		if(_childs.Count()>0){
			Json* childs = js->GetProperty(L"childs",1);
			for(uint i=0;i<_childs.Count();i++){
				_childs[i].ToJson(childs->AddArrayElm());
			}
		}
	}
	bool Layout::FromJson(Json* js){
		Clear();
		LayoutParam param;
		if(!js->GetPropertyInt32(L"style",param.Style)) return false;
		double v = 1.0;
		if(!js->GetPropertyDouble(L"weight",v)) return false;
		param.Weight = (float)v;
		if(!js->GetPropertyInt32(L"length",param.Length)) return false;
		if(!js->GetPropertyInt32(L"margin_left",param.Margin.left)) return false;
		if(!js->GetPropertyInt32(L"margin_top",param.Margin.top)) return false;
		if(!js->GetPropertyInt32(L"margin_right",param.Margin.right)) return false;
		if(!js->GetPropertyInt32(L"margin_bottom",param.Margin.bottom)) return false;
		LPCWSTR ctrlClass = js->PropertyStr(L"ctrl_class");
		if(WcsLength(ctrlClass)!=0){
			if(_createWnd){
				_createWnd->Close();
			}else if(!_createWnd){
				_createWnd = CreateCtrlObject(ctrlClass);
			}
			_createWnd->Param->ClassName = ctrlClass;
			_createWnd->Param->Text = js->PropertyStr(L"ctrl_text");
			js->GetPropertyInt32(L"ctrl_style",(int&)_createWnd->Param->Style);
			js->GetPropertyInt32(L"ctrl_exstyle",(int&)_createWnd->Param->ExStyle);
			int id = 0;
			if(js->GetPropertyInt32(L"ctrl_id",id))
				_createWnd->Param->Identity = id;
		}else{
			if(_createWnd){
				_createWnd->Close();
				delete _createWnd;
				_createWnd = 0;
			}
		}
		_wnd = 0;
		Param = param;
		Name = js->PropertyStr(L"name");
		Json* childs = js->GetProperty(L"childs");
		if(childs){
			for(childs->ArrayFirst();;childs->ArrayMove(1)){
				Json* child = childs->GetArrayElm();
				if(child==0) break;
				Layout* newLay = new Layout();
				if(!newLay->FromJson(child)){
					delete newLay;
					continue;
				}
				AddIn(newLay);
			}
		}
		return true;
	}
	void Layout::Reset(){
		Param.Reset();
		if(_createWnd){
			_createWnd->Close();
			delete _createWnd;
			_createWnd = 0;
		}
		Name = L"";
		_childs.Clear();
	}
	int Layout::_setChildHor(int left,int top,int width,int height){
		float sumWeight = 0;
		int sumDim = 0;
		for(uint i=0;i<_childs.Count();i++){
			LayoutParam& lp = _childs[i].Param;
			sumWeight += lp.Weight;
			sumDim += lp.Length + lp.Margin.left + lp.Margin.right;
		}
		int rest = width - sumDim;
		float x = (float)left;
		for(uint i=0;i<_childs.Count();i++){
			Layout& lay = _childs[i];
			LayoutParam& lp = lay.Param;
			float rationW;
			if(weightIsZero(sumWeight))
				rationW = (float)lp.Length;
			else
				rationW = lp.Length + rest*lp.Weight/sumWeight;
			x += lp.Margin.left;
			lay._setRect((int)x,top+lp.Margin.top,(int)(x+rationW)-(int)x,height-lp.Margin.top-lp.Margin.bottom);
			x += rationW + lp.Margin.right;
		}
		return sumDim;
	}
	int Layout::_setChildVer(int left,int top,int width,int height){
		float sumWeight = 0;
		int sumDim = 0;
		for(uint i=0;i<_childs.Count();i++){
			LayoutParam& lp = _childs[i].Param;
			sumWeight += lp.Weight;
			sumDim += lp.Length + lp.Margin.top + lp.Margin.bottom;
		}
		int rest = height - sumDim;
		float y = (float)top;
		for(uint i=0;i<_childs.Count();i++){
			Layout& lay = _childs[i];
			LayoutParam& lp = lay.Param;
			float rationH;
			if(weightIsZero(sumWeight))
				rationH = (float)lp.Length;
			else
				rationH = lp.Length + rest*lp.Weight/sumWeight;
			y += lp.Margin.top;
			lay._setRect(left+lp.Margin.left,(int)y,width-lp.Margin.left-lp.Margin.right,(int)(y+rationH)-(int)y/*消除舍1导致的一个像素的间距*/);
			y += rationH + lp.Margin.bottom;
		}
		return sumDim;
	}
	void Layout::_createControlTree(HWND parent){
		if(_createWnd!=0&&_wnd==0){
			_createWnd->Create(parent);
			_wnd = _createWnd->Handle();
		}
		for(uint i=0;i<_childs.Count();i++){
			_childs[i]._createControlTree(parent);
		}
	}
	void Layout::Refresh(HWND parent){
		if(parent) _createControlTree(parent);
		if(_parent){
			_parent->_setRect(_parent->_location.x,_parent->_location.y,_parent->_size.x,_parent->_size.y);
		}else{
			_setRect(_location.x,_location.y,_size.x,_size.y);
		}
	}
	bool Layout::Save(LPCWSTR file,DWORD code){
		Json json;
		ToJson(&json);
		return json.SaveToFile(file,code);
	}
	bool Layout::Load(LPCWSTR file,DWORD code){
		Json json;
		if(!json.LoadFromFile(file,code)) return false;
		Param.Reset();
		Name = L"";
		_childs.Clear();
		return FromJson(&json);
	}
	bool Layout::Load(ResID res,ResID rt,DWORD code){
		Param.Reset();
		Name = L"";
		_childs.Clear();
		DWORD size = 0;
		void* pm = GetResource(res,rt,&size);
		String str;
		str.FromMultiByte((LPCSTR)pm,size,code);
		Json json;
		if(!json.Parse(str)) return false;
		return FromJson(&json);
	}
	int Layout::Move(int pix){
		if(!_parent) return 0;
		uint index = _parent->IndexOf(this);
		if(index==0||index==(_parent->Count()-1)){
			//最前面和最后面的 Layout 无法调整位置。
			return 0;
		}

		Layout* preLay = _parent->Element(index-1);
		Layout* nextLay = _parent->Element(index+1);
		int dx1 = preLay->Param.Length + pix;
		int dx2 = nextLay->Param.Length - pix;
		int dx = Minimum(dx1,dx2);
		if(dx<0) pix += dx;

		if(pix==0) return 0;
		if(pix<0&&preLay->Param.Length<=0) return 0;
		if(pix>0&&nextLay->Param.Length<=0) return 0;

		preLay->Param.Length += pix;
		nextLay->Param.Length -= pix;
		_parent->_setRect(_parent->_location.x,_parent->_location.y,_parent->_size.x,_parent->_size.y);
		return pix;
	}
	Layout* Layout::GetElementByName(LPCWSTR name){
		if(Name==name) return this;
		for(uint i=0;i<_childs.Count();i++){
			Layout* lay = _childs[i].GetElementByName(name);
			if(lay) return lay;
		}
		return 0;
	}
	void TopLayout::_onSize(Message* msg,IWnd*){
		if(msg->wParam==SIZE_MINIMIZED) return;
		Twin16 cs = *((Twin16*)&msg->lParam);
		// Scroll 版本的
		int w = Maximum(_container->MinSize().x,(int)cs.x);
		int h = Maximum(_container->MinSize().y,(int)cs.y);
		_setRect(Param.Margin.left-_container->Scr().x,
			Param.Margin.top-_container->Scr().y,
			w-Param.Margin.left-Param.Margin.right,
			h-Param.Margin.top-Param.Margin.bottom);
		_container->Invalidate();
	}
	void TopLayout::Reset(){
		Layout::Reset();
		Name = L"Root";
		Param.Weight = 0;
		Param.Margin.SetValue(10,10,10,10);
	}
	TopLayout::TopLayout():_container(0){
		Name = L"Root";
		Param.Margin.SetValue(10,10,10,10);
		Param.Weight = 0;
	}
	void TopLayout::Refresh(HWND parent){
		Twin16 cs;
		if(!_container->GetClientSize(cs)) return;
		int w = Maximum(_container->MinSize().x,(int)cs.x);
		int h = Maximum(_container->MinSize().y,(int)cs.y);
		_location.SetValue(Param.Margin.left-_container->Scr().x,
			Param.Margin.top-_container->Scr().y);
		_size.SetValue(w-Param.Margin.left-Param.Margin.right,
			h-Param.Margin.top-Param.Margin.bottom);
		Layout::Refresh(*_container);
	}
	void TopLayout::SetContainer(IWnd* view){
		if(_container==view) return;
		if(_container!=NULL){
			_container->GetMsgHook(WM_SIZE)->Delete(this,&TopLayout::_onSize);
		}
		_container = view;
		if(!view) return;
		view->GetMsgHook(WM_SIZE)->Add(this,&TopLayout::_onSize);
		Refresh();
	}
	/*DragLayout::DragLayout():_dragLay(0){
	}
	void DragLayout::SetContainer(IWnd* view){
		if(_container==view) return;
		if(_container!=NULL){
			_container->GetMsgHook(WM_SIZE)->Delete(this,&DragLayout::_onSize);
			_container->OnMouse.Delete(this,&DragLayout::_onMouse);
		}
		_container = view;
		if(!view) return;
		view->GetMsgHook(WM_SIZE)->Add(this,&DragLayout::_onSize);
		view->OnMouse.Add(this,&DragLayout::_onMouse);
		Refresh();
	}
	bool ptInLayoutAndCanDrag(int x,int y,Layout* lay){
		if(!lay->Param.CanDrag()) return false;
		Twin16 lc = lay->GetLocation();
		Twin16 sz = lay->GetSize();
		Rect rect(lc.x,lc.y,lc.x+sz.x,lc.y+sz.y);
		return rect.PtInRect(x,y);
	}
	Layout* getDragLayout(int x,int y,Layout* lay){
		int count = lay->Count() - 1;
		for(int i=1;i<count;i++){
			Layout* child = lay->Element(i);
			if(ptInLayoutAndCanDrag(x,y,child)) return child;
		}
		count++;
		for(int i=0;i<count;i++){
			Layout* child = lay->Element(i);
			child = getDragLayout(x,y,child);
			if(child) return child;
		}
		return NULL;
	}
	void DragLayout::_calcPixelPerWeight(Layout* parent){
		float sumWeight = 0;
		int sumDim = 0;
		for(uint i=0;i<parent->Count();i++){
			LayoutParam& lp = parent->Element(i)->Param;
			sumWeight += lp.Weight;
			sumDim += lp.Length + lp.Margin.left + lp.Margin.right;
		}
		int rest = parent->GetSize().x - sumDim;
		_pixPerWeight = rest/sumWeight;
	}
	void DragLayout::_onMouse(Mouse* m){
		if(m->Event==WM_LBUTTONDOWN){
			_dragLay = getDragLayout(m->X,m->Y,this);
			if(_dragLay==0) return;
			int index = _dragLay->GetParent()->IndexOf(_dragLay);
			_preLay = _dragLay->GetParent()->Element(index-1);
			_nextLay = _dragLay->GetParent()->Element(index+1);
			Twin16 lc = _dragLay->GetLocation();
			_x = m->X - lc.x;
			_y = m->Y - lc.y;
			_calcPixelPerWeight(_dragLay->GetParent());
			::SetCapture(*_container);
			if(!OnDrag.IsNull()) OnDrag(_dragLay);
		}else if(m->Event==WM_LBUTTONUP){
			if(_dragLay==0) return;
			_dragLay = 0;
			::ReleaseCapture();
		}else if(m->Event==WM_MOUSEMOVE){
			if(_dragLay==0){
				return;
			}
			Layout* parent = _dragLay->GetParent();
			int dl;
			if(parent->Param.IsHorizontal()){
				int x = m->X - _dragLay->GetLocation().x;
				dl = x - _x;
			}else{
				int y = m->Y - _dragLay->GetLocation().y;
				dl = y - _y;
				//PrintD(L"dl:%d,y:%d,_y:%d,lc.y:%d",dl,y,_y,_dragLay->GetLocation().y);
			}
			//if(dl==0) return;
			//required changed weight.
			float dw = dl/_pixPerWeight;
			float npw = _preLay->Param.Weight + dw;
			float nnw = _nextLay->Param.Weight - dw;
			if(npw<0||nnw<0) return;
			_preLay->Param.Weight = npw;
			_nextLay->Param.Weight = nnw;
			parent->Refresh(*_container);
			if(!OnDrag.IsNull()) OnDrag(_dragLay);
		}
	}*/
}

File: jsuser_1.6/jsres/layout.js
function LayoutParam() {
    this.margin = new Rect(0,0,0,0);
    this.weight = 1;
    this.style = 0;
    this.length = 0;
}
LayoutParam.prototype.setVertical = function () {
    this.style |= 1;
    this.style -= 1;
}
LayoutParam.prototype.setHorizontal = function () {
    this.style |= 1;
}
LayoutParam.prototype.isHorizontal = function () {
    return (this.style & 1)==1;
}
LayoutParam.prototype.isVertical = function () {
    return (this.style & 1)==0;
}
LayoutParam.prototype.reset = function () {
    this.margin.setValue(0,0,0,0);
    this.length = 0;
    this.weight = 1;
	this.style = 0;
}

function Layout(){
	this._location = new Point();
	this._size = new Size();
    this.param = new LayoutParam();
	this.name = "";
	this.control = null;
	this.childs = [];
}
Layout.prototype._setRect = function(left,top,width,height){
	this._location.x = left;
	this._location.y = top;
	this._size.x = width;
	this._size.y = height;
	if(this.control instanceof CWnd){
		this.control.setPosition(left,top);
		this.control.setSize(width,height);
	}
	if(this.param.isHorizontal())
		this._setChildHor(left,top,width,height);
	else
		this._setChildVer(left,top,width,height);
}
Layout.prototype._weightIsZero = function(weight){
	return ((weight<0.0001)&&(weight>-0.0001));
}
Layout.prototype.getByName = function(name){
	if(this.name===name) return this;
	var count = this.childs.length;
	for(var i=0;i<count;i++){
		var lay = this.childs[i].getByName(name);
		if(lay) return lay;
	}
	return undefined;
}
Layout.prototype._setChildHor = function(left,top,width,height){
	var sumWeight = 0;
	var sumDim = 0;
	var count = this.childs.length;
	for(var i=0;i<count;i++){
		var lp = this.childs[i].param;
		sumWeight += lp.weight;
		sumDim += lp.length + lp.margin.left + lp.margin.right;
	}
	var rest = width - sumDim;
	var x = left;
	var count = this.childs.length;
	for(var i=0;i<count;i++){
		var lay = this.childs[i];
		var lp = lay.param;
		var rationW;
		if(this._weightIsZero(sumWeight))
			rationW = lp.length;
		else
			rationW = lp.length + rest*lp.weight/sumWeight;
		x += lp.margin.left;
		lay._setRect(Math.floor(x),top+lp.margin.top,Math.floor(x+rationW)-Math.floor(x),height-lp.margin.top-lp.margin.bottom);
		x += rationW + lp.margin.right;
	}
	return sumDim;
}
Layout.prototype._setChildVer = function(left,top,width,height){
	var sumWeight = 0;
	var sumDim = 0;
	var count = this.childs.length;
	for(var i=0;i<count;i++){
		var lp = this.childs[i].param;
		sumWeight += lp.weight;
		sumDim += lp.length + lp.margin.top + lp.margin.bottom;
	}
	var rest = height - sumDim;
	var y = top;
	var count = this.childs.length;
	for(var i=0;i<count;i++){
		var lay = this.childs[i];
		var lp = lay.param;
		var rationH;
		if(this._weightIsZero(sumWeight))
			rationH = lp.length;
		else
			rationH = lp.length + rest*lp.weight/sumWeight;
		y += lp.margin.top;
		lay._setRect(left+lp.margin.left,Math.floor(y),width-lp.margin.left-lp.margin.right,Math.floor(y+rationH)-Math.floor(y)/*消除舍1导致的一个像素的间距*/);
		y += rationH + lp.margin.bottom;
	}
	return sumDim;
}
Layout.prototype._calcMinSize = function(lay,parentHor){
	var sz = new Size(0,0);
	if(parentHor){
		sz.x = lay.param.length;
	}else{
		sz.y = lay.param.length;
	}
	var ms = new Size(0,0);
	var count = lay.childs.length;
	for(var i=0;i<count;i++){
		var child = lay.childs[i];
		var childSz = this._calcMinSize(child,lay.param.isHorizontal());
		ms.x += childSz.x;
		ms.y += childSz.y;
	}
	if(sz.x<ms.x) sz.x = ms.x;
	if(sz.y<ms.y) sz.y = ms.y;

	sz.x += lay.param.margin.left + lay.param.margin.right;
	sz.y += lay.param.margin.top + lay.param.margin.bottom;
	return sz;
}
Layout.prototype.ptInLayout = function(x,y){
	var rect = new Rect(this._location.x,this._location.y,this._location.x+this._size.x,this._location.y+this._size.y);
	return rect.ptInRect(x,y);
}
Layout.prototype.reset = function(){
	var count = this.childs.length;
	for(var i=0;i<count;i++){
		var child = this.childs[i];
		child.reset();
	}
	this.param.reset();
	if(this.control instanceof CWnd){
		this.control.dispose();
	}
	this.control = null;
	this.name = "";
	this.childs = [];
}
Layout.prototype.parent = function(){
	return _parent;
}
Layout.prototype.setControl = function(className,id,text,style,exstyle){
	if(className){
		if(this.control){
			this.control.close();
		}else{
			this.control = this._createCtrlObject(className);
		}
		this.control.className = className;
		this.control.text = text;
		this.control.id = id;
		this.control.style = style;
		this.control.exStyle = exStyle;
	}else if(this.control){
		this.control.dispose();
		this.control = null;
	}
}
Layout.prototype._createControlTree = function(hwnd){
	if(!hwnd) return;
	if(this.control){
		this._createControl(hwnd);
	}
	var count = this.childs.length;
	for(var i=0;i<count;i++){
		this.childs[i]._createControlTree(hwnd);
	}
}
Layout.prototype.refresh = function(hwnd){
	if(hwnd){
		this._createControlTree(hwnd);
	}
	var _parent = this._parent;
	if(_parent){
		_parent._setRect(_parent._location.x,_parent._location.y,_parent._size.x,_parent._size.y);
	}else{
		this._setRect(this._location.x,this._location.y,this._size.x,this._size.y);
	}
}
Layout.prototype.move = function(pix){
	if(!this._parent) return 0;
	var index = this._parent.childs.indexOf(this);
	if(index==0||index==(this._parent.childs.length-1)){
		//最前面和最后面的 Layout 无法调整位置。
		return 0;
	}

	var preLay = this._parent.childs[index-1];
	var nextLay = this._parent.childs[index+1];
	var dx1 = preLay.Param.Length + pix;
	var dx2 = nextLay.Param.Length - pix;
	var dx = Math.min(dx1,dx2);
	if(dx<0) pix += dx;

	if(pix==0) return 0;
	if(pix<0&&preLay.param.length<=0) return 0;
	if(pix>0&&nextLay.param.length<=0) return 0;

	preLay.param.length += pix;
	nextLay.param.length -= pix;
	this._parent._setRect(this._parent._location.x,this._parent._location.y,this._parent._size.x,this._parent._size.y);
	return pix;
}
Layout.prototype.add = function(index){
	var lay = new Layout();
	lay._parent = this;
	if(isNaN(index)) index = this.childs.length;
	this.childs.splice(index,0,lay);
	return lay;
}
Layout.prototype.del = function(index){
	if(isNaN(index)) return undefined;
	if(index>=this.childs.length||index!=-1) return undefined;
	var lay = this.childs.splice(index,1);
	if(!lay) return undefined;
	lay = lay[0];
	lay.reset();
	return true;
}
Layout.prototype.remove = function(index){
	if(isNaN(index)) return undefined;
	if(index>=this.childs.length||index!=-1) return undefined;
	return this.childs.splice(index,1);
}
Layout.prototype.addin = function(lay,index){
	lay._parent = this;
	if(isNaN(index)||index<0||index>this.childs.length) index = this.childs.length;
	this.childs.splice(index,0,lay);
	return lay;
}
Layout.prototype.toJson = function(){
	var json = {};
	json.name = this.name;
	json.style = this.param.style;
	json.weight = this.param.weight;
	json.length = this.param.length;
	json.margin_left = this.param.margin.left;
	json.margin_top = this.param.margin.top;
	json.margin_right = this.param.margin.right;
	json.margin_bottom = this.param.margin.bottom;
	if(this.control){
		var ctrl = this.control;
		json.ctrl_class = ctrl.className;
		json.ctrl_text = ctrl.text;
		json.ctrl_style = ctrl.style;
		json.ctrl_exstyle = ctrl.exStyle;
		json.ctrl_id = ctrl.id;
	}
	var count = this.childs.length;
	if(count>0) json.childs = [];
	for(var i=0;i<count;i++){
		var childjson = this.childs[i].toJson();
		json.childs.push(childjson);
	}
	return json;
}
Layout.prototype.parseJson = function(json){
	this.reset();
	var param = this.param;
	param.style = json.style;
	param.weight = json.weight;
	param.length = json.length;
	param.margin.setValue(json.margin_left,json.margin_top,json.margin_right,json.margin_bottom);
	if(typeof json.ctrl_class == "string"){
		this.control = this._createCtrlObject(json.ctrl_class);
		var cp = this.control;
		cp.text = json.ctrl_text;
		cp.style = json.ctrl_style;
		cp.exStyle = json.ctrl_exstyle;
		cp.id = json.id;
	}
	this.name = json.name;
	if(json.childs instanceof Array){
		var count = json.childs.length;
		for(var i=0;i<count;i++){
			var child = new Layout();
			child.parseJson(json.childs[i]);
			this.addin(child);
		}
	}
}
Layout.prototype.load = function(jsonfile,code){
	var file = new CFile();
	if(!file.create(jsonfile,true)) return false;
	if(!code) code = 65001;
	var jstr = file.readString(code);
	file.dispose();
	var json = JSON.parse(jstr);
	if(!json) return false;
	this.parseJson(json);
	return true;
}
Layout.prototype.save = function(jsonfile,code){
	var file = new CFile();
	if(!file.create(jsonfile)) return false;
	file.length = 0;
	var json = this.toJson();
	var str = JSON.stringify(json,undefined,'\t');
	if(!code) code = 65001;
	file.writeString(str,code);
	file.dispose();
	return true;
}




















